src/java/org/apache/cassandra/utils/memory/MemtableCleanerThread.java

```diff
@@ -27,61 +33,118 @@
+
+    public static class Clean<P extends MemtablePool> implements InterruptibleRunnable
     {
         /** The pool we're cleaning */
         final P pool;
 
         /** should ensure that at least some memory has been marked reclaiming after completion */
-        final Runnable cleaner;
+        final MemtableCleaner cleaner;
 
         /** signalled whenever needsCleaning() may return true */
         final WaitQueue wait = new WaitQueue();
 
-        private Clean(P pool, Runnable cleaner)
+        /** This is incremented when a cleaner is invoked and decremented when a cleaner has completed */
+        final AtomicInteger numPendingTasks = new AtomicInteger(0);
+
+        /** The maximum number of pending cleaning tasks, beyond this limit we pass the cleaner the threshold so that the cleaner
+         * will not clean unless it has more than the threshold to clean. */
+        final int maxPendingTasks;
+
+        /** The cleaning threshold, when we have more pending tasks than {@link this#maxPendingTasks}, then we pass this
+         * threshold to the cleaner as a minimum memory to clean *.
+         */
+        final double cleanThreshold;
+
+        private Clean(P pool, MemtableCleaner cleaner, int maxPendingTasks, double cleanThreshold)
         {
             this.pool = pool;
             this.cleaner = cleaner;
+            this.maxPendingTasks = maxPendingTasks;
+            this.cleanThreshold = cleanThreshold;
         }
 
-        boolean needsCleaning()
+        /** Return the number of pending tasks */
+        public int numPendingTasks()
         {
-            return pool.onHeap.needsCleaning() || pool.offHeap.needsCleaning();
+            return this.numPendingTasks.get();
         }
 
         @Override
         public void run() throws InterruptedException
         {
-            if (needsCleaning())
-            {
-                cleaner.run();
-            }
-            else
+            if (!pool.needsCleaning())
             {
                 final WaitQueue.Signal signal = wait.register();
-                if (!needsCleaning())
+                if (!pool.needsCleaning())
                     signal.await();
                 else
                     signal.cancel();
             }
+            else
+            {
+                // Increment the number of pending tasks and determine the minimum ownership ratio to pass to the clener
+                // This is zero (no min) if we have less than maxPendingTasks, otherwise it is the cleanThreshold
+                // The reason is that we don't want to queue more than numPendingTasks, but at the same time we don't
+                // want to end up with memtables that are too large, so if there is a memtable larger than cleanThreshold,
+                // we want to flush that
+                int numPendingTasks = this.numPendingTasks.incrementAndGet();
+                double minOwnershipRatio = numPendingTasks > maxPendingTasks ? cleanThreshold : 0;
```

Comment:
@adelapena @ekaterinadimitrova2 I think I'm approaching the end of my first pass of review, and the big thing bothering me at this point is that, on the one hand, we seem to have fixed a major potential source of unnecessary flush task generation (making sure we count writes between setting the discarding state and the actual flush as reclaiming), but on the other, we have introduced the complexity of additional configuration and parameterized execution of the cleaner.

In other words, if we're now properly informing `SubPool#updateNextClean()`, it seems like it shouldn't generate tasks when the cleaning threshold isn't reached anyway. We could simplify things by just removing the limit on pending tasks, but if we want to make sure that we can identify a similar problem in the future (i.e. a bug, not something we expect) we could also just have a simpler accounting of the number of pending tasks and expose it as a metric so operators can see what's happening (and we can debug more easily).

Thoughts?

Reply:
Excuse me for the delay, back to this today. As this came as part of a few different customer issues I had to follow up. Confirmed, the issue was completely solved so we can really skip the parameter and improve with a metric as you suggested. I am working on this tonight. Thank you!